slidechain/watch.go

```diff
@@ -117,62 +118,153 @@ func (c *Custodian) watchExports(ctx context.Context) {
-					AssetXDR []byte `json:"asset"`
-					Temp     string `json:"temp"`
-					Seqnum   int64  `json:"seqnum"`
-					Exporter string `json:"exporter"`
-				}
-				err := json.Unmarshal(infoItem[2].(txvm.Bytes), &info)
-				if err != nil {
-					continue
-				}
+			logItem := tx.Log[1]
+			var info struct {
+				AssetXDR []byte `json:"asset"`
+				Temp     string `json:"temp"`
+				Seqnum   int64  `json:"seqnum"`
+				Exporter string `json:"exporter"`
+				Amount   int64  `json:"amount"`
+				Anchor   []byte `json:"anchor"`
+				Pubkey   []byte `json:"pubkey"`
+			}
+			err := json.Unmarshal(i10rjson.HexBytes(logItem[2].(txvm.Bytes)), &info)
+			if err != nil {
+				continue
+			}
+			exportedAssetBytes := txvm.AssetID(importIssuanceSeed[:], info.AssetXDR)
 
-				// Check this Stellar asset code corresponds to retiredAssetIDBytes.
-				gotAssetID32 := txvm.AssetID(importIssuanceSeed[:], info.AssetXDR)
-				if !bytes.Equal(gotAssetID32[:], retiredAssetIDBytes) {
-					continue
-				}
+			// Record the export in the db,
+			// then wake up a goroutine that executes peg-outs on the main chain.
+			const q = `
+				INSERT INTO exports 
+				(txid, exporter, amount, asset_xdr, temp, seqnum, anchor, pubkey)
+				VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`
+			_, err = c.DB.ExecContext(ctx, q, tx.ID.Bytes(), info.Exporter, info.Amount, info.AssetXDR, info.Temp, info.Seqnum, info.Anchor, info.Pubkey)
+			if err != nil {
+				log.Fatalf("recording export tx: %s", err)
+			}
 
-				var exporter xdr.AccountId
-				err = exporter.SetAddress(info.Exporter)
-				if err != nil {
-					continue
-				}
+			log.Printf("recorded export: %d of txvm asset %x (Stellar %x) for %s", info.Amount, exportedAssetBytes, info.AssetXDR, info.Exporter)
 
-				// Record the export in the db,
-				// then wake up a goroutine that executes peg-outs on the main chain.
-				const q = `
-					INSERT INTO exports 
-					(txid, exporter, amount, asset_xdr, temp, seqnum)
-					VALUES ($1, $2, $3, $4, $5, $6)`
-				_, err = c.DB.ExecContext(ctx, q, tx.ID.Bytes(), exporter.Address(), retiredAmount, info.AssetXDR, info.Temp, info.Seqnum)
-				if err != nil {
-					log.Fatalf("recording export tx: %s", err)
-				}
+			c.exports.Broadcast()
+		}
+	}
+}
 
-				log.Printf("recorded export: %d of txvm asset %x (Stellar %x) for %s", retiredAmount, retiredAssetIDBytes, info.AssetXDR, exporter.Address())
+// Runs as a goroutine
+func (c *Custodian) watchPegOuts(ctx context.Context) {
+	defer log.Print("watchPegOuts exiting")
 
-				c.exports.Broadcast()
+	// Poll the database every minute for unprocessed exports.
+	go func(ctx context.Context) {
+		defer log.Print("export table poll exiting")
+		for t := time.Tick(1 * time.Minute); ; {
+			// Lock access to the export database.
+			// We do this to avoid double-processing any peg-out tx, via selecting it in both go-routines.
+			// We unlock at the end of an iteration of this go-routine or before any early return.
+			c.exportmux.Lock()
```

Comment:
When using a mutex, try to organize the code so that you can `defer` a call to `Unlock` as soon as you `Lock` it, and then return from that function as soon as possible to release the lock.

One thing you definitely want to avoid is trying to `Unlock` the mutex "manually" in every possible exit path from the function. (Plus, to be thorough, you'd need a `recover` function to handle non-local exits, so you can't really avoid `defer`.)